physical/redis.go

```diff
@@ -0,0 +1,314 @@
+	_, err := conn.Do("SET", realKey, encoded)
+	return err
+}
+
+func (c *RedisBackend) Get(key string) (*Entry, error) {
+	conn := c.pool.Get()
+	defer conn.Close()
+
+	reply, err := conn.Do("GET", fmt.Sprintf("%s/data/%s", c.path, key))
+	if err != nil {
+		return nil, err
+	}
+	if reply == nil {
+		// No value at this key, so don't try to decode.
+		return nil, nil
+	}
+
+	replyStr, err := redis.String(reply, err)
+	if err != nil {
+		return nil, err
+	}
+
+	value, err := base64.StdEncoding.DecodeString(replyStr)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Entry{
+		Key:   key,
+		Value: value,
+	}, nil
+}
+
+func (c *RedisBackend) Delete(key string) error {
+	conn := c.pool.Get()
+	defer conn.Close()
+
+	_, err := conn.Do("DEL", fmt.Sprintf("%s/data/%s", c.path, key))
+	return err
+}
+
+func (c *RedisBackend) List(prefix string) ([]string, error) {
+	conn := c.pool.Get()
+	defer conn.Close()
+
+	// Construct a "directory"-style name with a trailing slash from
+	// `prefix`.  But`prefix` may be the empty string, so be careful
+	// how we add "/".
+	realPrefix := fmt.Sprintf("%s/data/%s", c.path, prefix)
+	if !strings.HasSuffix(realPrefix, "/") {
+		realPrefix += "/"
+	}
+
+	// Ask Redis to list all keys beneath our prefix.  Note that this
+	// is not terribly efficient if you have a lot of keys, but we hope
+	// it's not a common operating.  There are more complex APIs for
+	// doing this incrementally.
+	reply, err := conn.Do("KEYS", realPrefix+"*")
+	matches, err := redis.Strings(reply, err)
+
+	// TODO - Don't recurse subdirectories.  This means stripping
+	// everything after "/" and removing duplicates.
+	results := make([]string, 0)
+	dirs := make(map[string]bool)
+	for _, match := range matches {
+		// Remove the path that we were querying from the results.
+		match = strings.TrimPrefix(match, realPrefix)
+
+		slashPos := strings.Index(match, "/")
+		if slashPos == -1 {
+			// We have an ordinary file, so return it directly.
+			results = append(results, match)
+		} else {
+			// Keep the trailing slash.
+			dir := match[0 : slashPos+1]
+
+			// Have we seen this directory already?  If not,
+			// return it and mark it as seen.
+			if _, present := dirs[dir]; !present {
+				results = append(results, dir)
+				dirs[dir] = true
+			}
+		}
+	}
+
+	return results, err
+}
+
+func (c *RedisBackend) LockWith(key, value string) (Lock, error) {
+	return &RedisLock{
+		key:    fmt.Sprintf("%s/_lock/%s", c.path, key),
+		value:  value,
+		pool:   c.pool,
+		logger: c.logger,
+	}, nil
+}
+
+type RedisLock struct {
+	key    string
+	value  string
+	pool   *redis.Pool
+	logger *log.Logger
+}
+
+// Acquire the lock.  To interrupt lock acquistion, close stopCh.  The
+// returned channel will be closed if the lock is lost.
+func (c *RedisLock) Lock(stopCh <-chan struct{}) (<-chan struct{}, error) {
+	// Loop until we get a non-nil reply from Redis indicating a
+	// successful operation.
+	var reply interface{}
```

Comment:
This is how the other back ends work, AFAICT. It's better to follow the existing code than to get clever. :-)